src/atomic.rs

```diff
@@ -1,160 +1,380 @@
+        Ok(Self {
+            nonce: 0,
+            input_ids: TCid::new_hamt(bs)?,
+            exec_ids: TCid::new_hamt(bs)?,
+        })
     }
 
-    #[allow(dead_code)]
-    pub fn from_raw_bytes(ser: &RawBytes) -> anyhow::Result<Self> {
-        Ok(cbor::deserialize_params(ser)?)
+    pub fn load(cid: &Cid, bs: &impl Blockstore) -> anyhow::Result<Option<AtomicExecRegistry>> {
+        bs.get_cbor(cid)
     }
-}
 
-/// State of an actor including a lock to support atomic executions.
-#[derive(Serialize, Deserialize_tuple)]
-pub struct LockableState<T>
-where
-    T: Serialize + DeserializeOwned + MergeableState<T>,
-{
-    lock: bool,
-    state: T,
+    pub fn init_atomic_exec<'a, S>(
+        &mut self,
+        bs: &impl Blockstore,
+        state: impl IntoIterator<Item = &'a mut S>,
+        input: AtomicInput,
+        lock: bool,
+    ) -> anyhow::Result<AtomicInputID>
+    where
+        S: LockableState + 'a,
+    {
+        // Optionally lock the state and compute its CIDs
+        let unlocked_state_cids = state.into_iter().try_fold(Vec::new(), |mut v, s| {
+            if lock {
+                s.lock()?;
+            } else {
+                v.push(s.cid());
+            }
+            anyhow::Ok(v)
+        })?;
+
+        // Generate and register a new input ID
+        let input_id = self.new_input_id(&unlocked_state_cids, &input);
+        self.input_ids.modify(bs, |m| {
+            let k = BytesKey::from(input_id.bytes());
+            let v = m.set(
+                k,
+                AtomicInputEntry {
+                    unlocked_state_cids,
+                    input,
+                },
+            )?;
+            assert!(v.is_some(), "input ID collision");
+            Ok(())
+        })?;
+
+        Ok(input_id)
+    }
+
+    pub fn cancel_atomic_exec<'a, S>(
+        &mut self,
+        bs: &impl Blockstore,
+        input_id: AtomicInputID,
+        input_fn: impl FnOnce(AtomicInput) -> Box<dyn Iterator<Item = &'a mut S>>,
+    ) -> anyhow::Result<()>
+    where
+        S: LockableState + 'a,
+    {
+        // Consume own input ID and retrieve the associated data
+        let AtomicInputEntry { input, .. } = self.input_ids.modify(bs, |m| {
+            let k = BytesKey::from(input_id.bytes());
+            let (_, v) = m
+                .delete(&k)?
+                .ok_or_else(|| anyhow::anyhow!("unexpected input ID"))?;
+            Ok(v)
+        })?;
+
+        // Get the state and ensure it's unlocked
+        let state_iter = input_fn(input);
+        state_iter.for_each(|s| {
+            if s.is_locked() {
+                s.unlock().unwrap();
+            }
+        });
+
+        Ok(())
+    }
+
+    pub fn prepare_atomic_exec<'a, S, I>(
+        &mut self,
+        bs: &impl Blockstore,
+        own_input_id: AtomicInputID,
+        input_ids: &HashMap<IPCAddress, AtomicInputID>,
+        input_fn: impl FnOnce(AtomicInput) -> (I, Box<dyn Iterator<Item = &'a mut S>>),
```

Comment:
We need a better description of what is the role of `input_fn` and `output_fn`, it may make more sense to me once you have the actor example, but so far I don't see how the `output` can be computed here or the role of `input_fn`.

I guess `output_fn` is the atomic execution we want to perform, right? But these should probably be described as a list of messages to execute.

Reply:
I added more comments which may help to better understand the role of those closures.

The output is computed from the optional input data associated with actor's own input ID and any values captured by the closure from its environment. The input data may include the messages to execute, whereas the closure's environment would include any additional data from other executing actors that is required to produce the output.

Reply:
Ok, it makes more sense now. I'll wait for unit tests or an example actor to be ready to the final pass on this matter :) Thanks!